src/services/reader/web_sys.rs

```diff
@@ -0,0 +1,160 @@
+//! `web-sys` implementation for the reader service.
+
+use crate::callback::Callback;
+use crate::services::Task;
+use gloo::events::EventListener;
+use js_sys::Uint8Array;
+use std::cmp;
+use std::fmt;
+pub use web_sys::{Blob, File};
+use web_sys::{Event, FileReader};
+
+/// Struct that represents data of a file.
+#[derive(Clone, Debug)]
+pub struct FileData {
+    /// Name of loaded file.
+    pub name: String,
+    /// Content of loaded file.
+    pub content: Vec<u8>,
+}
+
+/// Struct that represents a chunk of a file.
+#[derive(Clone, Debug)]
+pub enum FileChunk {
+    /// Reading of chunks started. Equals **0%** progress.
+    Started {
+        /// Name of loaded file.
+        name: String,
+    },
+    /// The next data chunk that read. Also provides a progress value.
+    DataChunk {
+        /// The chunk of binary data.
+        data: Vec<u8>,
+        /// The progress value in interval: `0 < progress <= 1`.
+        progress: f32,
+    },
+    /// Reading of chunks finished. Equals **100%** progress.
+    Finished,
+}
+
+/// A reader service attached to a user context.
+#[derive(Default, Debug)]
+pub struct ReaderService {}
+
+impl ReaderService {
+    /// Creates a new service instance connected to `App` by provided `sender`.
+    pub fn new() -> Self {
+        Self {}
+    }
+
+    /// Reads all bytes from a file and returns them with a callback.
+    pub fn read_file(&mut self, file: File, callback: Callback<FileData>) -> ReaderTask {
+        let file_reader = FileReader::new().unwrap();
+        let reader = file_reader.clone();
+        let name = file.name();
+        let callback = move |_event: &Event| {
+            let array = Uint8Array::new_with_byte_offset(
+                &reader
+                    .result()
+                    .expect("`FileReader` hasn't finished loading"),
```

Comment:
The result can also indicate if the read succeeded or not so this is a common path, should not panic 😬 

>  The value is null if the reading is not yet complete or was unsuccessful.
https://developer.mozilla.org/en-US/docs/Web/API/FileReader/result

I think the result should be bubbled up to the user through the callback. It's ok to deviate from `stdweb` behaviour. How about `Callback<Option<FileData>>`?

Reply:
In `stdweb` we didn't handle this, we just did nothing instead, if we decide to bubble this up I suggest to do it in `stdweb` too.
I changed it to mirror the `stdweb` behaviour now, do nothing if there is no result.

Reply:
It's ok if the API's are different between `web-sys` and `stdweb`.

Reply:
Alright, as far as [MDN](https://developer.mozilla.org/en-US/docs/Web/API/FileReader/result) says:
> The value is null if the reading is not yet complete or was unsuccessful.

Considering that we use `loadend`, it will only fail if it's unsuccessful.
So I moved the codepath to continue reading more chunks into success, so it won't try if it's a failure.

> It's ok if the API's are different between `web-sys` and `stdweb`.

I mirrored it, it was pretty straightforward.